### ISSUE TITLE
Database helper shouldn't contain MySQL-specific sql

### ISSUE THREAD

**Issue by SuchAFuriousDeath** [espocrm/espocrm#2546]
**Is your feature request related to a problem? Please describe.**
A bit of a follow-up on #2537 ( I'm trying to make EspoCRM work with Postgres). In the class Espo\Core\Utils\Database, calling the methods 'doesSupportFulltext' and 'getMaxIndexLength' causes the SQL in the 'getTableEngine' method to be run, causing an error (Postgres has no such thing as engines, so the SQL fails). 

**Describe the solution you'd like**
Factor the platform-specific SQL into another class, perhaps make the current helper abstract and make child classes for each platform.

**Describe alternatives you've considered**
The Helper can be probably replaced the same way you suggested in the previous issue (#2537). I just don't consider it very nice and thought you might want to take a look at this, since you fullfilled my previous feature request.

**Comment by SuchAFuriousDeath** [comment]
Also, if one wanted to add a Platform and Driver for Postgres, the helper would have to be overridden in the same inconvenient way.

**Comment by yurikuzn** [comment]
I made some changes, supposed to work with `database.platform => 'Potsgresql'` now. But I didn't test. Hardcoded w/o introducing a new interface.

**Comment by SuchAFuriousDeath** [comment]
@yurikuzn Now it throws a new error. I believe it's these 2 lines: 

https://github.com/espocrm/espocrm/blob/ac667356c528d0764e78dc6406dd09498deb6a1d/application/Espo/Core/Utils/Database/Helper.php#L140-L142

Reading the Driver code, it looks like it wants the version in the format returned by the method 'getDatabaseVersion' (for example 13.8.0) as opposed to the full version.

**Comment by yurikuzn** [comment]
Not an easy fix. As it's opposite for MySQL and MardiaDB.

**Comment by SuchAFuriousDeath** [comment]
Isn't it just a condition?

**Comment by yurikuzn** [comment]
Please try `getDatabaseVersion` and let me know. I will be able to look not today.

**Comment by yurikuzn** [comment]
It's too much of hardcode. Eventually, it should be refactory.

**Comment by SuchAFuriousDeath** [comment]
@yurikuzn 
Right, I understand. Before your commit regarding the first issue, I already got further with the module. I got to the point where more than 100 tables were created (Now I'm writing the module from scratch based on the new 7.4 commit). From what I can remember, these are the issues I came across:

- **Reserved keywords**: Postgres has many reserved keywords, but among them are the words 'user' and 'group'. Someone could easily create an Entity named group and there already is an entity called User. This means that basically all identifiers have to be quoted, including table and column names.
- **I believe that unquoted strings are treated as case-insensitive**, which means that all select aliases have to be quoted, otherwise something like "SELECT created_by AS createdBy", would in PHP be stored under the key 'createdby' instead.
- _note_: Postgres quoting is done with double quotes (") 
- **Different type names**. Postgres doesn't have LONGTEXT and TINYINT, it has only TEXT and BOOLEAN. The boolean values have to be passed as true and false instead of 1 and 0. This can be partially solved by overriding the 'quote' function in the QueryComposer, but I think i saw it hardcoded as 0 somewhere.
- **Index collisions** - This is quite something. All index names have to be unique across the whole database, not just one table. They could also collide with other table names (and also views, but Espo doesn't use those), but that probably won't happen, but I'm putting it here just for the sake of completeness
If I remember any other issues, I'll put them here.

**Comment by yurikuzn** [comment]
Tables and aliases are quoted in base-composer. What is not quoted is column identifiers like `user.name`. Should they also be quoted?

**Comment by SuchAFuriousDeath** [comment]
Unfortunately, they do.

**Comment by SuchAFuriousDeath** [comment]
I managed to find almost all the places (Somewhere it's hardcoded as not quoted), but I couldn't manage to some of them quoted. I remember that of them was related to userData.

**Comment by yurikuzn** [comment]
There're should not be any raw SQL in Espo except DBAL. Expression is used.

**Comment by SuchAFuriousDeath** [comment]
Espo\Core\Utils\Database\Orm\Fields\PersonName is the only one I came across that was problematic so far. In the getSelect method, there is a 'ISNULL'.

**Comment by yurikuzn** [comment]
These are not SQL. They are in complex-expression syntax, a special syntax I created for Espo.

**Comment by SuchAFuriousDeath** [comment]
I know, i probably misspoke

**Comment by SuchAFuriousDeath** [comment]
Also, i have no idea about fulltext search. Didn't really get that far. Might be completely incompatible.

**Comment by yurikuzn** [comment]
Fulltext search is one of few items where Postgres may beat MySQL, as in MySQL full-text indexes can't be used together with other indexes (needed when applying order) that makes it very slow.

**Comment by SuchAFuriousDeath** [comment]
Just remembered about collations and added them to the list.

**Comment by SuchAFuriousDeath** [comment]
> Please try `getDatabaseVersion` and let me know. I will be able to look not today.

@yurikuzn I just tried and it works.

**Comment by SuchAFuriousDeath** [comment]
@yurikuzn 
By the way, I tried the newest commit and rebuilding throws the following error.

`
Failed to rebuild database schema. Details: SQLSTATE[42601]: Syntax error: 7 ERROR:  syntax error at or near "$1" LINE 1: SHOW $1 at /home/postgres.devcrm.cz/espocrm-git/application/Espo/Core/Utils/Database/DBAL/Factories/PostgresqlConnectionFactory.php:84 [] []
`

I googled it and I suspect that Postgres doesn't support prepared statements with SHOW, but I'm not sure

**Comment by SuchAFuriousDeath** [comment]
@yurikuzn Tried the new new commit, and the CREATE TABLE queries throw and error because of the nonexistent collation.

**Comment by yurikuzn** [comment]
I can't look into this issue anymore. Try to solve occurred problems by hardcoding.

**Comment by yurikuzn** [comment]
Fixing it in an iteration fashion is not the way to go. I don't have an environment, the issue is not in our backlog, it will keep me distracted from the planned work. If you fix major issues by yourself hardcoded, then I will be able to look into and see the whole picture.

**Comment by boydle** [comment]
I have run in to problems with table size and join limits in Maria DB and Mysql these do not appear to be a problem with Postgress. 

Support for postgress would be a great enhancement, @SuchAFuriousDeath if you manage to implement this will you be releasing your work back to the espocrm project or in an extension or package?

**Comment by yurikuzn** [comment]
As I mentioned above, we unlikely be supporting Postgres in the nearest future. It will add us a lot of extra work, for our support, for developers. MySQL and MariaDB took years to polish. It may seem easy if you have 1 client, it's a pain if you have 1000nds users.

**Comment by SuchAFuriousDeath** [comment]
I plan to release it publicly on my github once I get it somewhat working. But even if I do, I'll make sure to include a warning that using it with Postgres is highly experimental and shouldn't be relied on if you're a company. It's just a side project of mine, I'm still a student and have a job, so I shouldn't be relied on to fix possible issues very quickly. I noticed that some people have already asked about Postgres and I thought it would be an interesting challenge. All bug reports and such will be welcome and I'd gladly cooperate with Yuri to put it in Espo officially, if he chooses to maintain it sometime in the future.

**Comment by SuchAFuriousDeath** [comment]
Just did a pretty long session of debugging. I got it to a point where it lets you in (authentication is successful). Also, another item to add to the list is the FIELD function which doesn't exist in Postgres and will have to be implemented through the case statement, probably.

**Comment by SuchAFuriousDeath** [comment]
Found another thing - when activities are selected using the Activities Service, sometimes an empty string is selected with an alias like 'dateStartDate' and it throws and error. Not quite sure why.

**Comment by SuchAFuriousDeath** [comment]
Haven't figured it out yet, but I found out something else. Index hinting is completely non-existent as a feature.

**Comment by SuchAFuriousDeath** [comment]
I figured it out, I think. Postgres seems to be more strict with types, so when you select '' as dateStartDate in query of the union and date_start_date (a column with type date), these two clash, because an empty string is not a valid date value.. So, maybe convert the empty string to a null value in the queryComposer?

**Comment by yurikuzn** [comment]
> convert the empty string to a null value in the queryComposer?

I think to convert to null in expressions.
